File: src/c++/dynamic_programming/LIS/1818/main.cc
/* Date : 2019.5.11
 * Author : seungyeon choi
 * 1818 (책정리)
 */

#include <algorithm>
#include <cstdio>

using namespace std;

const int MAX_COUNT = 200001;

int n, x, ans = 0;
int DP[MAX_COUNT];

int main() {
  scanf("%d", &n);

  for (int i = 0; i < n; i++) {
    scanf("%d", &x);
    if (DP[ans] < x)
      DP[++ans] = x;
    else
      *lower_bound(DP, DP + ans + 1, x) = x;
  }

  printf("%d", n - ans);

  return 0;
}

File: src/python/dynamic_programming/LIS/1818/main.py
# Date : 2017.11.16
# Author : seungyeon choi
# 1818 (책정리)

N = int(input())
S = list(map(int, input().split()))
DP = []
DP.append(S[0])
for i in range(1, N):
    if S[i] > DP[len(DP)-1] :
        DP.append(S[i])
    else :
        def BS(a, b, v):
            while a < b :
                mid = (a + b) // 2
                if DP[mid] >= v : b = mid
                else : a = mid + 1                    
            return b
        DP[BS(0, len(DP)-1, S[i])] = S[i]
print (N-len(DP))
